web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect, useRef } from 'react';
+import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
+
+import {
+  ResourcesResponse,
+  ResourceFilter,
+  resourceFilterToHookDeps,
+} from 'teleport/services/agents';
+import { UrlResourcesParams } from 'teleport/config';
+
+/**
+ * Supports fetching more data from the server when more data is available. Pass
+ * a `fetchFunc` that retrieves a single batch of data. After the initial
+ * request, the server is expected to return a `startKey` field that denotes the
+ * next `startKey` to use for the next request.
+ *
+ * This hook is intended to be used in tandem with the `useInfiniteScroll` hook.
+ * See its documentation for an example.
```

Comment:
If this hook is only used in tandem with `useInfiniteScroll`, then how about we just call it from within `useInfiniteScroll` itself? So `useInfiniteScroll` gets all of its props, and also can return all the values that this currently returns which might need to be used elsewhere. Like

```ts
export function useInfiniteScroll(props: KeyBasedPaginationProps & UseInfiniteScrollProps) {
   const { fetch, forceFetch, resources, attempt } = useKeyBasedPagination({...});
   
  // use fetch in here, rather than needing to pass it in as a prop   
  
  return { forceFetch, resources, attempt };
}
```

Reply:
Because this way, it's easier to test the `useInfiniteScroll` in isolation. But you got me thinking here; perhaps I could just mock out the `useKeyBasedPagination` hook in the tests instead. Or even the resource fetcher itself.